src/caffe/util/im2col.cpp

```diff
@@ -10,9 +10,12 @@ void im2col_cpu(const Dtype* data_im, const int channels,
     const int height, const int width, const int kernel_h, const int kernel_w,
     const int pad_h, const int pad_w,
     const int stride_h, const int stride_w,
+    const int dilation_h, const int dilation_w,
     Dtype* data_col) {
-  const int height_col = (height + 2 * pad_h - kernel_h) / stride_h + 1;
-  const int width_col = (width + 2 * pad_w - kernel_w) / stride_w + 1;
+  const int height_col = (height + 2 * pad_h -
+      (dilation_h * (kernel_h - 1)  + 1)) / stride_h + 1;
+  const int width_col = (width + 2 * pad_w -
+      (dilation_w * (kernel_w - 1)  + 1)) / stride_w + 1;
```

Comment:
Hi guys,
Thanks for including me in the loop.
Dilated convolution is just another name for convolution with holes (aka
atrous convolution). In my initial caffe code I was using the name "hole"
for the corresponding parameter. I have since been using the name
"input_stride" which I believe is more descriptive and appropriate (and
will also be used in the forthcoming release of this feature in
TensorFlow). Dilation is a term that also describes a nonlinear filtering
operation in mathematical morphology and I find it confusing to use in this
context. I thus vote for "input_stride".
On Dec 16, 2015 2:19 AM, "Fisher Yu" <notifications@github.com> wrote:

> In src/caffe/util/im2col.cpp
> <https://github.com/BVLC/caffe/pull/3452#discussion_r47737695>:
>
> >      Dtype* data_col) {
> > -  const int height_col = (height + 2 * pad_h - kernel_h) / stride_h + 1;
> > -  const int width_col = (width + 2 * pad_w - kernel_w) / stride_w + 1;
> > +  const int height_col = (height + 2 * pad_h -
> > +      (dilation_h * (kernel_h - 1)  + 1)) / stride_h + 1;
> > +  const int width_col = (width + 2 * pad_w -
> > +      (dilation_w * (kernel_w - 1)  + 1)) / stride_w + 1;
>
> Sorry that I don't quite understand this comment. Which space is extra?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/BVLC/caffe/pull/3452/files#r47737695>.
>